ARCs/arc-0054.md

```diff
@@ -83,142 +93,120 @@ switch create_NoOp NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENT
-	frame_dig -2 // asa: asset
+	frame_dig -1 // asa: asset
 	asset_holding_get AssetBalance
 	swap
 	pop
 	!
 	assert
 
-	// contracts/asa-burn-app.algo.ts:15
-	// assert(asa.clawback === globals.zeroAddress)
-	frame_dig -2 // asa: asset
-	asset_params_get AssetClawback
-	assert
-	global ZeroAddress
-	==
-	assert
-
-	// contracts/asa-burn-app.algo.ts:17
-	// preMBR = globals.currentApplicationAddress.minBalance
-	global CurrentApplicationAddress
-	acct_params_get AcctMinBalance
-	assert
-	frame_bury -3 // preMBR: uint64
-
-	// contracts/asa-burn-app.algo.ts:19
+	// contracts/arc54.algo.ts:15
 	// sendAssetTransfer({
 	//       assetReceiver: globals.currentApplicationAddress,
 	//       xferAsset: asa,
 	//       assetAmount: 0,
+	//       fee: 0,
 	//     })
 	itxn_begin
 	int axfer
 	itxn_field TypeEnum
 
-	// contracts/asa-burn-app.algo.ts:20
+	// contracts/arc54.algo.ts:16
 	// assetReceiver: globals.currentApplicationAddress
 	global CurrentApplicationAddress
 	itxn_field AssetReceiver
 
-	// contracts/asa-burn-app.algo.ts:21
+	// contracts/arc54.algo.ts:17
 	// xferAsset: asa
-	frame_dig -2 // asa: asset
+	frame_dig -1 // asa: asset
 	itxn_field XferAsset
 
-	// contracts/asa-burn-app.algo.ts:22
+	// contracts/arc54.algo.ts:18
 	// assetAmount: 0
 	int 0
 	itxn_field AssetAmount
 
-	// Fee field not set, defaulting to 0
+	// contracts/arc54.algo.ts:19
+	// fee: 0
 	int 0
 	itxn_field Fee
 
 	// Submit inner transaction
 	itxn_submit
-
-	// contracts/asa-burn-app.algo.ts:25
-	// verifyTxn(mbrPayment, {
-	//       receiver: globals.currentApplicationAddress,
-	//       amount: { greaterThanEqualTo: globals.currentApplicationAddress.minBalance - preMBR },
-	//     })
-	// verify receiver
-	frame_dig -1 // mbrPayment: pay
-	gtxns Receiver
-	global CurrentApplicationAddress
-	==
-	assert
-
-	// verify amount
-	frame_dig -1 // mbrPayment: pay
-	gtxns Amount
-	global CurrentApplicationAddress
-	acct_params_get AcctMinBalance
-	assert
-	frame_dig -3 // preMBR: uint64
-	-
-	>=
-	assert
 	retsub
 
-abi_route_defaultTEALScriptCreate:
+abi_route_createApplication:
 	int 1
 	return
 
 create_NoOp:
-	txn NumAppArgs
-	bz abi_route_defaultTEALScriptCreate
+	method "createApplication()void"
+	txna ApplicationArgs 0
+	match abi_route_createApplication
 	err
 
 call_NoOp:
-	method "optIntoASA(pay,asset)void"
+	method "optIntoASA(asset)void"
 	txna ApplicationArgs 0
 	match abi_route_optIntoASA
 	err
 ```
 
+### TealScript Source Code
+```
+import { Contract } from '@algorandfoundation/tealscript';
+
+// eslint-disable-next-line no-unused-vars
+class ARC54 extends Contract {
+  /*
+   * Sends an inner transaction to opt the contract account into an ASA.
+   * Will fail if the contract is already opted in to the asset.
+   * The fee for the inner transaction must be covered by the caller.
+   *
+   * @param asa The ASA to opt in to
+   */
+  optIntoASA(asa: Asset): void {
+    assert(!globals.currentApplicationAddress.hasAsset(asa));
```

Comment:
I'd remove this assert.  Its simply friction for callers - with this they have to know/check in advance whether the burn contract has already 'burned' the specified asset.  It should always work, opted-in or not.  If it's already opted-in, it's a no-op, return success - whether the caller paid MBR is irrelevant.  Once they send to the burn contract account, it still succeeds which is all they care about.

Reply:
Good suggestion--I have made this change.

Reply:
Good spot @pbennett . The optin state may change between the client reading the optins and the txn hitting the chain, which could cause contract calls to fail unnecessarily. Given the 0 fee enforcement, this is safe to do